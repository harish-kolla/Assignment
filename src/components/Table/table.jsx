import React, { useState, useEffect } from "react";
import Table from "react-bootstrap/Table";
import InfiniteScroll from "react-infinite-scroll-component";
import axios from "axios";

import { HeaderComponent } from "../Header/header";
import { AddNewRecord } from "../NewRecord/newRecord";
import { staticKeys } from "../../constants";

export const CustomTable = () => {
  const [originalData, updateOriginalData] = useState([]);
  const [items, updateItems] = useState([]);
  const [displayRow, updateRow] = useState(false);
  const [recordsCount, updateRecordsCount] = useState(50);
  const [hasMoreItems, updateMoreItems] = useState(true);
  const [newRecord, updateRecord] = useState("");

  useEffect(() => {
    getData();
  }, []);

  const getData = () => {
    axios.get("data.json").then((res) => {
      updateOriginalData(res?.data); //storing 200k records data in originalData state, So that when we remove record will not update complete list(wil update only items list)
      updateItems(res?.data?.slice(0, 50));
    });
  };

  const addNewRow = () => {
    scrollToTop();
    if (newRecord) {
      addRecordDetals();
    }
    updateRecord("");
    updateRow(true);
  };

  const addRecordDetals = () => {
    let newItems = [...items];
    newItems.unshift(newRecord);
    updateItems(newItems);
    updateRecord("");
    updateRow(false);
  };

  const inputChangeHandler = (e, field) => {
    let updatedNewRecord = { ...newRecord };
    updatedNewRecord[field] = e.target.value;
    updateRecord(updatedNewRecord);
  };

  const removeRecord = (index) => {
    let newItems = [...items];
    newItems.splice(index, 1);
    updateItems(newItems);
  };

  const fetchMoreRecords = () => {
    if (originalData.length <= recordsCount) {
      updateMoreItems(!hasMoreItems);
    }
    setTimeout(() => {
      updateRecordsCount(recordsCount + 50); // added setTimeout to delay loading response (as we are not triggering API call)
    }, 1000);
  };

  useEffect(() => {
    updateItems(originalData.slice(0, recordsCount)); //loading 50 records at a time, On scroll will load other records
  }, [recordsCount, originalData]);

  const scrollToTop = () => {
    window.scrollTo({ top: 0, left: 0, behavior: "smooth" });
  };

  return (
    <div>
      <HeaderComponent addNewRow={addNewRow} scrollToTop={scrollToTop} />
      <InfiniteScroll
        dataLength={recordsCount}
        next={fetchMoreRecords}
        hasMore={hasMoreItems}
        loader={<h4>{staticKeys.loadMore}</h4>}
      >
        <Table bordered hover>
          <thead>
            <tr>
              <th width="170">{staticKeys.item}</th>
              <th width="170">{staticKeys.itemDesc}</th>
              <th width="170">{staticKeys.price}</th>
              <th width="100">{staticKeys.delete}</th>
            </tr>
          </thead>
          <tbody>
            {displayRow && (
              <AddNewRecord
                inputChangeHandler={inputChangeHandler}
                newRecord={newRecord}
                updateRow={updateRow}
              />
            )}
            {items?.map((record, index) => (
              <tr data-testid="row" key={index}>
                <td data-testid={`item-${index}`}>{record.item}</td>
                <td>{record.itemDesc}</td>
                <td>{record.price}</td>
                <td>
                  <button
                    data-testid={`remove-${index}`}
                    type="button"
                    className="btn btn-link"
                    onClick={() => removeRecord(index)}
                  >
                    {staticKeys.remove}
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </InfiniteScroll>
    </div>
  );
};
