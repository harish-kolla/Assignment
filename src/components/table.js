import React, { useState, useEffect } from "react";
import Table from "react-bootstrap/Table";
import InfiniteScroll from "react-infinite-scroll-component";
import axios from "axios";
import { HeaderComponent } from "./header";
import { AddNewRecord } from "./newRecord";

const tableData = [
  {
    item: "item 1",
    itemDesc: "item",
    price: 200,
    id: 1,
  },
  {
    item: "item 2",
    itemDesc: "item",
    price: 200,
    id: 2,
  },
  {
    item: "item 3",
    itemDesc: "item",
    price: 200,
    id: 3,
  },
  {
    item: "item 4",
    itemDesc: "item",
    price: 200,
    id: 4,
  },
  {
    item: "item 5",
    itemDesc: "item",
    price: 200,
    id: 5,
  },
  {
    item: "item 6",
    itemDesc: "item",
    price: 200,
    id: 6,
  },
  {
    item: "item 7",
    itemDesc: "item",
    price: 200,
    id: 7,
  },
];

export const CustomTable = () => {
  const [items, updateItems] = useState(tableData);
  const [displayRow, updateRow] = useState(false);
  const [recordsCount, updateRecordsCount] = useState(50);
  const [hasMoreItems, updateMoreItems] = useState(true);
  const [newRecord, updateRecord] = useState("");

  const addNewRow = (e) => {
    scrollToTop();
    if (newRecord) {
      addRecordDetals();
    }
    updateRecord("");
    updateRow(true);
  };

  const addRecordDetals = () => {
    let newItems = [...items];
    newItems.unshift(newRecord);
    updateItems(newItems);
    updateRecord("");
    updateRow(false);
  };

  const inputChangeHandler = (e, field) => {
    let updatedNewRecord = { ...newRecord };
    updatedNewRecord[field] = e.target.value;
    updateRecord(updatedNewRecord);
  };

  const removeRecord = (index) => {
    let newItems = [...items];
    newItems.splice(index, 1);
    updateItems(newItems);
  };

  const fetchMoreRecords = () => {
    if (items.length <= recordsCount) {
      updateMoreItems(!hasMoreItems);
    }
    setTimeout(() => {
      updateRecordsCount(recordsCount + 50);
    }, 1000);
  };

  const scrollToTop = () => {
    window.scrollTo({ top: 0, left: 0, behavior: "smooth" });
  };

  return (
    <div>
      <HeaderComponent
        addNewRow={addNewRow}
        scrollToTop={scrollToTop}
        itemsCount={items?.length}
      />
      <InfiniteScroll
        dataLength={recordsCount}
        next={fetchMoreRecords}
        hasMore={hasMoreItems}
        loader={<h4>Loading 50 more records...</h4>}
      >
        <Table bordered hover>
          <thead>
            <tr>
              <th width="170">Item</th>
              <th width="170">Item Description</th>
              <th width="170">Price</th>
              <th width="100">Delete</th>
            </tr>
          </thead>
          <tbody>
            {displayRow && (
              <AddNewRecord
                inputChangeHandler={inputChangeHandler}
                newRecord={newRecord}
                updateRow={updateRow}
              />
            )}
            {items?.slice(0, recordsCount).map((record, index) => (
              <tr key={index}>
                <td>{record.item}</td>
                <td>{record.itemDesc}</td>
                <td>{record.price}</td>
                <td>
                  <button
                    type="button"
                    className="btn btn-link"
                    onClick={() => removeRecord(index)}
                  >
                    Remove
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </InfiniteScroll>
    </div>
  );
};
