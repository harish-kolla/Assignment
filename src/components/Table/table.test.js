import { render, fireEvent, act, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import axios from "axios";

import { CustomTable } from "./table";
import MockedRecords from "../../../public/data.json";

beforeEach(() => {
  axios.get.mockResolvedValueOnce({
    data: MockedRecords,
  });
});

afterEach(() => {
  axios.get.mockClear();
});

test("Snapshot of table", () => {
  const { container } = render(<CustomTable />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Should render Add Row button", () => {
  const { queryByText } = render(<CustomTable />);
  expect(queryByText("Add Row")).toBeTruthy();
});

test("Should render Top Page button", () => {
  const { queryByText } = render(<CustomTable />);
  expect(queryByText("To Top")).toBeTruthy();
});

test("Should render Infinite Scroll with `Loading 50 more records...` text", () => {
  const { queryByText } = render(<CustomTable />);
  expect(queryByText("Loading 50 more records...")).toBeTruthy();
});

test("Should render Table with `Item`,`Item Description`,`Price` and `Delete` column", () => {
  const { queryByText } = render(<CustomTable />);
  expect(queryByText("Item")).toBeTruthy();
  expect(queryByText("Item Description")).toBeTruthy();
  expect(queryByText("Price")).toBeTruthy();
  expect(queryByText("Delete")).toBeTruthy();
});

test("Should render Add New Row component when click on Add Row button", () => {
  const { getByText, getByTestId } = render(<CustomTable />);
  fireEvent.click(getByText("Add Row"));
  expect(getByTestId("add-new-row")).toBeTruthy();
});

test("Should render row based on given data", async () => {
  const { getAllByTestId, queryByTestId } = render(<CustomTable />);
  const rowValues = await waitFor(() =>
    getAllByTestId("row").map((row, index) => {
      return queryByTestId(`item-${index}`).textContent;
    })
  );
  expect(rowValues).toEqual(
    MockedRecords.slice(0, 50).map((data) => data.item)
  );
  expect(axios.get).toHaveBeenCalledWith("data.json");
  expect(axios.get).toHaveBeenCalledTimes(1);
});

test("Should remove row on click of remove button", async () => {
  const { getAllByTestId, queryByTestId } = render(<CustomTable />);
  await waitFor(() => getAllByTestId("row"));
  act(() => {
    fireEvent.click(queryByTestId("remove-1"));
  });
  const rowValues = await waitFor(() =>
    getAllByTestId("row").map((row, index) => {
      return queryByTestId(`item-${index}`).textContent;
    })
  );
  expect(rowValues.length).toEqual(50);
});

test("Should call inputChangeHandler on change new row input", () => {
  const { queryByTestId, getByText, getByDisplayValue } = render(
    <CustomTable />
  );
  fireEvent.click(getByText("Add Row"));
  fireEvent.change(queryByTestId("add-new-price"), { target: { value: "23" } });
  expect(getByDisplayValue("23")).toEqual(queryByTestId("add-new-price"));
});

test("Should call addNewRow on click Add Row button", async () => {
  const { queryByTestId, getByText, getByDisplayValue, getByTestId } = render(
    <CustomTable />
  );
  fireEvent.click(getByText("Add Row"));
  fireEvent.change(queryByTestId("add-new-price"), { target: { value: "23" } });
  expect(getByDisplayValue("23")).toEqual(queryByTestId("add-new-price"));
  fireEvent.click(getByText("Add Row"));
  expect(getByTestId("add-new-row")).toBeTruthy();
});
