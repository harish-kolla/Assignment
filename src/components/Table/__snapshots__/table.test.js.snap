// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of table 1`] = `
<div>
  <nav
    class="navbar sticky-top navbar-light "
    style="background: rgb(227, 242, 253);"
  >
    <button
      class="add-buttton btn btn-primary"
      data-testid="add-row-button"
      type="button"
    >
      Add Row
    </button>
    <span
      class="virtulized-list"
      data-testid="virtulized-list"
    >
      Virtualized List
    </span>
    <button
      class="btn btn-link"
      data-testid="scroll-top-button"
      type="button"
    >
      Top Page
    </button>
  </nav>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <table
        class="table table-bordered table-hover"
      >
        <thead>
          <tr>
            <th
              width="170"
            >
              Item
            </th>
            <th
              width="170"
            >
              Item Description
            </th>
            <th
              width="170"
            >
              Price
            </th>
            <th
              width="100"
            >
              Delete
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <h4>
        Loading 50 more records...
      </h4>
    </div>
  </div>
</div>
`;
